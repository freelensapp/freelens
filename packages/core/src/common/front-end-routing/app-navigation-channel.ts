/**
 * Copyright (c) Freelens Authors. All rights reserved.
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */

import type { MessageChannel } from "@freelensapp/messaging";
import { IpcRendererNavigationEvents } from "../ipc/navigation-events";

export type AppNavigationChannel = MessageChannel<string>;

export const appNavigationChannel: AppNavigationChannel = {
  id: IpcRendererNavigationEvents.NAVIGATE_IN_APP,
};
