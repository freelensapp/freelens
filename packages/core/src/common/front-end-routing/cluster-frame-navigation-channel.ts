/**
 * Copyright (c) Freelens Authors. All rights reserved.
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */

import type { MessageChannel } from "@freelensapp/messaging";
import { IpcRendererNavigationEvents } from "../ipc/navigation-events";

export type ClusterFrameNavigationChannel = MessageChannel<string>;

export const clusterFrameNavigationChannel: ClusterFrameNavigationChannel = {
  id: IpcRendererNavigationEvents.NAVIGATE_IN_CLUSTER,
};
