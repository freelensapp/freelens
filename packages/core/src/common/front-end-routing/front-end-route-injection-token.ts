/**
 * Copyright (c) Freelens Authors. All rights reserved.
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */

import { getInjectionToken } from "@ogre-tools/injectable";
import type { IComputedValue } from "mobx";
import type { LensRendererExtension } from "../../extensions/lens-renderer-extension";

export const frontEndRouteInjectionToken = getInjectionToken<Route<unknown>>({
  id: "front-end-route-injection-token",
});

export interface Route<TParameter = void> {
  path: string;
  clusterFrame: boolean;
  isEnabled: IComputedValue<boolean>;
  extension?: LensRendererExtension;

  readonly parameterSignature?: TParameter;
}
