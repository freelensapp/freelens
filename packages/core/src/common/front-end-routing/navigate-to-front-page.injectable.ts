/**
 * Copyright (c) Freelens Authors. All rights reserved.
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */

import { getInjectable } from "@ogre-tools/injectable";
import navigateToCatalogInjectable from "./routes/catalog/navigate-to-catalog.injectable";

const navigateToFrontPageInjectable = getInjectable({
  id: "navigate-to-front-page",
  instantiate: (di) => di.inject(navigateToCatalogInjectable),
});

export default navigateToFrontPageInjectable;
