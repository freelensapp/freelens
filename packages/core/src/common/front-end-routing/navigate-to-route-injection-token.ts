/**
 * Copyright (c) Freelens Authors. All rights reserved.
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */

import { getInjectionToken } from "@ogre-tools/injectable";
import type { Route } from "./front-end-route-injection-token";

type InferParametersFrom<TRoute> = TRoute extends Route<infer TParameters> ? TParameters : never;

type RequiredKeys<T> = Exclude<
  {
    [K in keyof T]: T extends Record<K, T[K]> ? K : never;
  }[keyof T],
  undefined
>;

type ObjectContainingNoRequired<T> = T extends void ? never : RequiredKeys<T> extends [] ? any : never;

type ObjectContainsNoRequired<T> = T extends ObjectContainingNoRequired<T> ? true : false;

// TODO: Missing types for:
// - Navigating to route without parameters, with parameters
// - Navigating to route with required parameters, without parameters
type Parameters<TParameters> = TParameters extends void
  ? {}
  : ObjectContainsNoRequired<TParameters> extends true
    ? { parameters?: TParameters }
    : { parameters: TParameters };

export type NavigateToRouteOptions<TRoute> = Parameters<InferParametersFrom<TRoute>> & {
  query?: Record<string, string>;
  fragment?: string;
  withoutAffectingBackButton?: boolean;
};

export type NavigateToRoute = <TRoute extends Route<unknown>>(
  route: TRoute,
  options?: NavigateToRouteOptions<TRoute>,
) => void;

export const navigateToRouteInjectionToken = getInjectionToken<NavigateToRoute>({
  id: "navigate-to-route-injection-token",
});
