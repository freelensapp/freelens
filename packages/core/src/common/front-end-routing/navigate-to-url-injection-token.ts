/**
 * Copyright (c) Freelens Authors. All rights reserved.
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */

import { getInjectionToken } from "@ogre-tools/injectable";

export interface NavigateToUrlOptions {
  withoutAffectingBackButton?: boolean;
  forceRootFrame?: boolean;
}

export type NavigateToUrl = (url: string, options?: NavigateToUrlOptions) => void;

export const navigateToUrlInjectionToken = getInjectionToken<NavigateToUrl>({ id: "navigate-to-url-injection-token" });
