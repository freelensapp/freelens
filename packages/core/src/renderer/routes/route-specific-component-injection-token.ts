/**
 * Copyright (c) Freelens Authors. All rights reserved.
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */

import { getInjectionToken } from "@ogre-tools/injectable";
import type { Route } from "../../common/front-end-routing/front-end-route-injection-token";

export const routeSpecificComponentInjectionToken = getInjectionToken<{
  route: Route<unknown>;
  Component: React.ElementType<any>;
}>({
  id: "route-specific-component-injection-token",
});
