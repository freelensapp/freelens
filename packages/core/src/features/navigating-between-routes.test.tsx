/**
 * Copyright (c) Freelens Authors. All rights reserved.
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */

import type { DiContainer } from "@ogre-tools/injectable";
import { getInjectable } from "@ogre-tools/injectable";
import type { RenderResult } from "@testing-library/react";
import { computed, runInAction } from "mobx";
import { observer } from "mobx-react";
import React from "react";
import type { Route } from "../common/front-end-routing/front-end-route-injection-token";
import { frontEndRouteInjectionToken } from "../common/front-end-routing/front-end-route-injection-token";
import { navigateToRouteInjectionToken } from "../common/front-end-routing/navigate-to-route-injection-token";
import type { ApplicationBuilder } from "../renderer/components/test-utils/get-application-builder";
import { getApplicationBuilder } from "../renderer/components/test-utils/get-application-builder";
import currentPathInjectable from "../renderer/routes/current-path.injectable";
import currentRouteInjectable from "../renderer/routes/current-route.injectable";
import queryParametersInjectable from "../renderer/routes/query-parameters.injectable";
import routePathParametersInjectable from "../renderer/routes/route-path-parameters.injectable";
import { routeSpecificComponentInjectionToken } from "../renderer/routes/route-specific-component-injection-token";

describe("navigating between routes", () => {
  let rendered: RenderResult;
  let builder: ApplicationBuilder;

  beforeEach(async () => {
    builder = getApplicationBuilder();
  });

  describe("given route without path parameters", () => {
    let windowDi: DiContainer;

    beforeEach(async () => {
      builder.beforeWindowStart(({ windowDi }) => {
        runInAction(() => {
          windowDi.register(testRouteWithoutPathParametersInjectable);
          windowDi.register(testRouteWithoutPathParametersComponentInjectable);
        });
      });

      rendered = await builder.render();

      windowDi = builder.applicationWindow.only.di;
    });

    describe("when navigating to route", () => {
      let route: Route;

      beforeEach(() => {
        const navigateToRoute = windowDi.inject(navigateToRouteInjectionToken);

        route = windowDi.inject(testRouteWithoutPathParametersInjectable);

        navigateToRoute(route);
      });

      it("renders", () => {
        expect(rendered.container).toMatchSnapshot();
      });

      it("knows current route", () => {
        const currentRoute = windowDi.inject(currentRouteInjectable);

        expect(currentRoute.get()).toBe(route);
      });

      it("knows current path", () => {
        const currentPath = windowDi.inject(currentPathInjectable);

        expect(currentPath.get()).toBe("/some-path");
      });

      it("does not have query parameters", () => {
        const queryParameters = windowDi.inject(queryParametersInjectable);

        expect(queryParameters.get()).toEqual({});
      });

      it("does not have path parameters", () => {
        const pathParameters = windowDi.inject(routePathParametersInjectable, route);

        expect(pathParameters.get()).toEqual({});
      });
    });

    it("when navigating to route with query parameters, knows query parameters", () => {
      const navigateToRoute = windowDi.inject(navigateToRouteInjectionToken);
      const queryParameters = windowDi.inject(queryParametersInjectable);

      const route = windowDi.inject(testRouteWithoutPathParametersInjectable);

      navigateToRoute(route, {
        query: {
          someParameter: "some-value",
          someOtherParameter: "some-other-value",
        },
      });

      expect(queryParameters.get()).toEqual({
        someParameter: "some-value",
        someOtherParameter: "some-other-value",
      });
    });
  });

  describe("given route with optional path parameters", () => {
    let windowDi: DiContainer;

    beforeEach(async () => {
      builder.beforeWindowStart(({ windowDi }) => {
        runInAction(() => {
          windowDi.register(routeWithOptionalPathParametersInjectable);
          windowDi.register(routeWithOptionalPathParametersComponentInjectable);
        });
      });

      rendered = await builder.render();

      windowDi = builder.applicationWindow.only.di;
    });

    describe("when navigating to route with path parameters", () => {
      let route: Route<any>;

      beforeEach(() => {
        route = windowDi.inject(routeWithOptionalPathParametersInjectable);

        const navigateToRoute = windowDi.inject(navigateToRouteInjectionToken);

        navigateToRoute(route, {
          parameters: {
            someParameter: "some-value",
            someOtherParameter: "some-other-value",
          },
        });
      });

      it("renders", () => {
        expect(rendered.container).toMatchSnapshot();
      });

      it("knows current route", () => {
        const currentRoute = windowDi.inject(currentRouteInjectable);

        expect(currentRoute.get()).toBe(route);
      });

      it("knows current path", () => {
        const currentPath = windowDi.inject(currentPathInjectable);

        expect(currentPath.get()).toBe("/some-path/some-value/some-other-value");
      });

      it("knows path parameters", () => {
        const pathParameters = windowDi.inject(routePathParametersInjectable, route);

        expect(pathParameters.get()).toEqual({
          someParameter: "some-value",
          someOtherParameter: "some-other-value",
        });
      });
    });

    describe("when navigating to route without path parameters", () => {
      let route: Route<any>;

      beforeEach(() => {
        route = windowDi.inject(routeWithOptionalPathParametersInjectable);

        const navigateToRoute = windowDi.inject(navigateToRouteInjectionToken);

        navigateToRoute(route);
      });

      it("knows current route", () => {
        const currentRoute = windowDi.inject(currentRouteInjectable);

        expect(currentRoute.get()).toBe(route);
      });

      it("knows current path", () => {
        const currentPath = windowDi.inject(currentPathInjectable);

        expect(currentPath.get()).toBe("/some-path");
      });

      it("knows path parameters", () => {
        const pathParameters = windowDi.inject(routePathParametersInjectable, route);

        expect(pathParameters.get()).toEqual({
          someParameter: undefined,
          someOtherParameter: undefined,
        });
      });
    });
  });
});

const testRouteWithoutPathParametersInjectable = getInjectable({
  id: "some-route",
  injectionToken: frontEndRouteInjectionToken,

  instantiate: () => ({
    path: "/some-path",
    clusterFrame: false,
    isEnabled: computed(() => true),
  }),
});

const testRouteWithoutPathParametersComponentInjectable = getInjectable({
  id: "some-route-component",

  instantiate: (di) => ({
    route: di.inject(testRouteWithoutPathParametersInjectable),
    Component: () => <div>Some component</div>,
  }),

  injectionToken: routeSpecificComponentInjectionToken,
});

const routeWithOptionalPathParametersInjectable = getInjectable({
  id: "some-route",
  injectionToken: frontEndRouteInjectionToken,

  instantiate: (): Route<{ someParameter?: string; someOtherParameter?: string }> => ({
    path: "/some-path/:someParameter?/:someOtherParameter?",
    clusterFrame: false,
    isEnabled: computed(() => true),
  }),
});

const routeWithOptionalPathParametersComponentInjectable = getInjectable({
  id: "some-route-component",

  instantiate: (di) => {
    const route = di.inject(routeWithOptionalPathParametersInjectable);
    const pathParameters = di.inject(routePathParametersInjectable, route);

    return {
      route,

      Component: observer(() => <pre>{JSON.stringify(pathParameters.get(), null, 2)}</pre>),
    };
  },

  injectionToken: routeSpecificComponentInjectionToken,
});
