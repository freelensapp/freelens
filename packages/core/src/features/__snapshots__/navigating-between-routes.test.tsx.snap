// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`navigating between routes given route with optional path parameters when navigating to route with path parameters renders 1`] = `
<div>
  <div
    class="Notifications flex column align-flex-end"
  />
  <div
    class="ClusterManager"
  >
    <div
      class="topBar"
    >
      <div
        class="items"
      >
        <div
          class="preventedDragging"
        >
          <i
            class="Icon material interactive focusable"
            data-testid="home-button"
            tabindex="0"
          >
            <span
              class="icon"
              data-icon-name="home"
            >
              home
            </span>
          </i>
        </div>
        <div
          class="size-sm"
        />
        <div
          class="preventedDragging"
        >
          <i
            class="Icon material interactive disabled focusable"
            data-testid="history-back"
          >
            <span
              class="icon"
              data-icon-name="arrow_back"
            >
              arrow_back
            </span>
          </i>
        </div>
        <div
          class="size-sm"
        />
        <div
          class="preventedDragging"
        >
          <i
            class="Icon material interactive disabled focusable"
            data-testid="history-forward"
          >
            <span
              class="icon"
              data-icon-name="arrow_forward"
            >
              arrow_forward
            </span>
          </i>
        </div>
        <div
          class="separator"
        />
      </div>
    </div>
    <main>
      <div
        id="lens-views"
      />
      <pre>
        {
  "someParameter": "some-value",
  "someOtherParameter": "some-other-value"
}
      </pre>
    </main>
    <div
      class="HotbarMenu flex column"
    >
      <div
        class="HotbarItems flex column gaps"
      >
        <div
          class="HotbarCell isDraggingOwner animateDown"
          index="0"
        >
          <div
            style="z-index: 12; position: absolute;"
          >
            <div
              class="HotbarIcon contextMenuAvailable"
            >
              <div
                class="Avatar rounded disabled avatar"
                id="hotbarIcon-hotbar-icon-catalog-entity"
                style="width: 40px; height: 40px; background: rgb(5, 1, 130);"
              >
                Ca
              </div>
            </div>
          </div>
        </div>
        <div
          class="HotbarCell isDraggingOwner animateDown"
          index="1"
        />
        <div
          class="HotbarCell isDraggingOwner animateDown"
          index="2"
        />
        <div
          class="HotbarCell isDraggingOwner animateDown"
          index="3"
        />
        <div
          class="HotbarCell isDraggingOwner animateDown"
          index="4"
        />
        <div
          class="HotbarCell isDraggingOwner animateDown"
          index="5"
        />
        <div
          class="HotbarCell isDraggingOwner animateDown"
          index="6"
        />
        <div
          class="HotbarCell isDraggingOwner animateDown"
          index="7"
        />
        <div
          class="HotbarCell isDraggingOwner animateDown"
          index="8"
        />
        <div
          class="HotbarCell isDraggingOwner animateDown"
          index="9"
        />
        <div
          class="HotbarCell isDraggingOwner animateDown"
          index="10"
        />
        <div
          class="HotbarCell isDraggingOwner animateDown"
          index="11"
        />
      </div>
      <div
        class="HotbarSelector"
      >
        <i
          class="Icon material interactive focusable"
          tabindex="0"
        >
          <span
            class="icon"
            data-icon-name="arrow_left"
          >
            arrow_left
          </span>
        </i>
        <div
          class="HotbarIndex"
        >
          <div
            class="badge Badge small clickable"
            data-testid="hotbar-menu-badge-for-Default"
            id="hotbarIndex"
          >
            1
          </div>
        </div>
        <i
          class="Icon material interactive focusable"
          tabindex="0"
        >
          <span
            class="icon"
            data-icon-name="arrow_right"
          >
            arrow_right
          </span>
        </i>
      </div>
    </div>
    <div
      class="StatusBar"
      data-testid="status-bar"
    >
      <div
        class="leftSide"
        data-testid="status-bar-left"
      />
      <div
        class="rightSide"
        data-testid="status-bar-right"
      />
    </div>
  </div>
</div>
`;

exports[`navigating between routes given route without path parameters when navigating to route renders 1`] = `
<div>
  <div
    class="Notifications flex column align-flex-end"
  />
  <div
    class="ClusterManager"
  >
    <div
      class="topBar"
    >
      <div
        class="items"
      >
        <div
          class="preventedDragging"
        >
          <i
            class="Icon material interactive focusable"
            data-testid="home-button"
            tabindex="0"
          >
            <span
              class="icon"
              data-icon-name="home"
            >
              home
            </span>
          </i>
        </div>
        <div
          class="size-sm"
        />
        <div
          class="preventedDragging"
        >
          <i
            class="Icon material interactive disabled focusable"
            data-testid="history-back"
          >
            <span
              class="icon"
              data-icon-name="arrow_back"
            >
              arrow_back
            </span>
          </i>
        </div>
        <div
          class="size-sm"
        />
        <div
          class="preventedDragging"
        >
          <i
            class="Icon material interactive disabled focusable"
            data-testid="history-forward"
          >
            <span
              class="icon"
              data-icon-name="arrow_forward"
            >
              arrow_forward
            </span>
          </i>
        </div>
        <div
          class="separator"
        />
      </div>
    </div>
    <main>
      <div
        id="lens-views"
      />
      <div>
        Some component
      </div>
    </main>
    <div
      class="HotbarMenu flex column"
    >
      <div
        class="HotbarItems flex column gaps"
      >
        <div
          class="HotbarCell isDraggingOwner animateDown"
          index="0"
        >
          <div
            style="z-index: 12; position: absolute;"
          >
            <div
              class="HotbarIcon contextMenuAvailable"
            >
              <div
                class="Avatar rounded disabled avatar"
                id="hotbarIcon-hotbar-icon-catalog-entity"
                style="width: 40px; height: 40px; background: rgb(5, 1, 130);"
              >
                Ca
              </div>
            </div>
          </div>
        </div>
        <div
          class="HotbarCell isDraggingOwner animateDown"
          index="1"
        />
        <div
          class="HotbarCell isDraggingOwner animateDown"
          index="2"
        />
        <div
          class="HotbarCell isDraggingOwner animateDown"
          index="3"
        />
        <div
          class="HotbarCell isDraggingOwner animateDown"
          index="4"
        />
        <div
          class="HotbarCell isDraggingOwner animateDown"
          index="5"
        />
        <div
          class="HotbarCell isDraggingOwner animateDown"
          index="6"
        />
        <div
          class="HotbarCell isDraggingOwner animateDown"
          index="7"
        />
        <div
          class="HotbarCell isDraggingOwner animateDown"
          index="8"
        />
        <div
          class="HotbarCell isDraggingOwner animateDown"
          index="9"
        />
        <div
          class="HotbarCell isDraggingOwner animateDown"
          index="10"
        />
        <div
          class="HotbarCell isDraggingOwner animateDown"
          index="11"
        />
      </div>
      <div
        class="HotbarSelector"
      >
        <i
          class="Icon material interactive focusable"
          tabindex="0"
        >
          <span
            class="icon"
            data-icon-name="arrow_left"
          >
            arrow_left
          </span>
        </i>
        <div
          class="HotbarIndex"
        >
          <div
            class="badge Badge small clickable"
            data-testid="hotbar-menu-badge-for-Default"
            id="hotbarIndex"
          >
            1
          </div>
        </div>
        <i
          class="Icon material interactive focusable"
          tabindex="0"
        >
          <span
            class="icon"
            data-icon-name="arrow_right"
          >
            arrow_right
          </span>
        </i>
      </div>
    </div>
    <div
      class="StatusBar"
      data-testid="status-bar"
    >
      <div
        class="leftSide"
        data-testid="status-bar-left"
      />
      <div
        class="rightSide"
        data-testid="status-bar-right"
      />
    </div>
  </div>
</div>
`;
